pkg/controller/service/servicecontroller.go

```diff
@@ -593,14 +593,57 @@ func hostsFromNodeSlice(nodes []*api.Node) []string {
 	return result
 }
 
+// findNodeRole returns the role of a given node, or "" if none found.
+// The role is determined by looking in order for:
+// * a kubernetes.io/role label
+// * a kubeadm.alpha.kubernetes.io/role label
+// * a taint with Key 'dedicated'
+// If no role is found, ("") is returned
+// TODO: promote this function into utils in api.  kubectl needs a similar function.
+func findNodeRole(node *api.Node) string {
+	if role := node.Labels["kubernetes.io/role"]; role != "" {
+		return role
+	}
+	if role := node.Labels["kubeadm.alpha.kubernetes.io/role"]; role != "" {
+		return role
+	}
+
+	taints, err := api.GetTaintsFromNodeAnnotations(node.Annotations)
+	if err != nil {
+		glog.Warningf("error parsing node taints for node %q: %v", node.Name, err)
+	} else {
+		for _, taint := range taints {
+			if taint.Key == "dedicated" {
```

Comment:
We should figure out what we are going to call this, ref https://github.com/kubernetes/kubernetes/pull/31647#issuecomment-2568505